src/engine/srv.c

```diff
@@ -214,6 +214,43 @@ dss_sleep(uint64_t msec)
 	return 0;
 }
 
+/**
+ * Wait for the eventual is ready until it is timeout.
+ *
+ * \param[in]	eventual	eventual to be waitted.
+ * \param[in]	status		eventual status.
+ * \param[in]	timeout		timeout seconds.
+ * \param[in]	step		sleep seconds if it is not ready.
+ *
+ * \return	status		if eventual is ready or failed within the timeout.
+ * 		DER_TIMEOUT	if it is not ready within timeout.
+ */
+int
+dss_eventual_timeout_wait(ABT_eventual eventual, void *status, uint32_t timeout, uint32_t step)
+{
+	ABT_bool is_ready;
+	uint64_t start = daos_gettime_coarse();
+	int	 rc;
+
+	while (1) {
+		rc = ABT_eventual_test(eventual, status, &is_ready);
+		if (rc != ABT_SUCCESS || is_ready == ABT_TRUE) {
+			if (rc != ABT_SUCCESS)
+				rc = dss_abterr2der(rc);
+			break;
+		}
+
+		if (daos_gettime_coarse() - start >= timeout) {
+			D_ERROR("Started from "DF_U64", timeout after %u\n", start, timeout);
+			rc = -DER_TIMEDOUT;
+			break;
+		}
+		dss_sleep(step);
```

Comment:
hm, why not use ABT_eventual_wait() and set up a timer that will signal the eventual upon timeout? Sounds like it would be more efficient wrt latency of waking up (common case) and also ABT scheduling, no?

Reply:
ah, yeah, this is probably better. I was dumb when I wrote this.

Reply:
ah, I remembered why I did not use timer here, because it needs to create another ULT or tasklet for the timer for all cases, which seems a bit expensive. 

Ideally, we only need create a timer for long delayed cases.  Hmm, we may check more frequently during initial few seconds, then extend the sleep time. hmm, may need discuss a bit.  @johannlombardi